plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java

```diff
@@ -627,11 +627,39 @@ private boolean reduceParameterizedTypeRef(ParameterizedTypeRef left, Parameteri
 						throw new UnsupportedOperationException("unsupported subtype of TypeArgument: "
 								+ leftArg.getClass().getName());
 					}
-					wasAdded |= reduce(leftArg, leftParamSubst, variance.mult(INV));
+					final TypeVariable leftTypeVar = (left.getDeclaredType() != null
+							&& left.getDeclaredType().getTypeVars().size() > 0)
+									? left.getDeclaredType().getTypeVars().get(0) : null;
+					final TypeVariable rightTypeVar = (right.getDeclaredType() != null
+							&& right.getDeclaredType().getTypeVars().size() >= idx)
+									? right.getDeclaredType().getTypeVars().get(idx) : null;
```

Comment:
there is a convenience method `Type#getVarianceOfTypeVar()` that might help simplify this a bit